[SPARK-37475][SQL] Add scale parameter to floor and ceil functions

sathiyapk opened the pull request:
<!--
Thanks for sending a pull request!  Here are some tips for you:
  1. If this is your first time, please read our contributor guidelines: https://spark.apache.org/contributing.html
  2. Ensure you have added or run the appropriate tests for your PR: https://spark.apache.org/developer-tools.html
  3. If the PR is unfinished, add '[WIP]' in your PR title, e.g., '[WIP][SPARK-XXXX] Your PR title ...'.
  4. Be sure to keep the PR description updated to reflect all changes.
  5. Please write your PR title to summarize what this PR proposes.
  6. If possible, provide a concise example to reproduce the issue for a faster review.
  7. If you want to add a new configuration, please read the guideline first for naming configurations in
     'core/src/main/scala/org/apache/spark/internal/config/ConfigEntry.scala'.
  8. If you want to add or modify an error type or message, please read the guideline first in
     'core/src/main/resources/error/README.md'.
-->

### What changes were proposed in this pull request?
<!--
Please clarify what changes you are proposing. The purpose of this section is to outline the changes and how this PR fixes the issue. 
If possible, please consider writing useful notes for better and faster reviews in your PR. See the examples below.
  1. If you refactor some codes with changing classes, showing the class hierarchy will help reviewers.
  2. If you fix some SQL features, you can provide some references of other DBMSes.
  3. If there is design documentation, please add the link.
  4. If there is a discussion in the mailing list, please add the link.
-->
Adds `scale` parameter to `floor`/`ceil` functions in order to allow users to control the rounding position. This feature is proposed in the PR: https://github.com/apache/spark/pull/34593

### Why are the changes needed?
<!--
Please clarify why the changes are needed. For instance,
  1. If you propose a new API, clarify the use case for a new API.
  2. If you fix a bug, you can clarify why it is a bug.
-->
Currently we support Decimal RoundingModes : HALF_UP (round) and HALF_EVEN (bround). But we have use cases that needs RoundingMode.UP and RoundingMode.DOWN.

Floor and Ceil functions helps to do this but it doesn't support the position of the rounding. Adding scale parameter to the functions would help us control the rounding positions. 

Snowflake supports `scale` parameter to `floor`/`ceil` :
` FLOOR( <input_expr> [, <scale_expr> ] )`

REF:
https://docs.snowflake.com/en/sql-reference/functions/floor.html

### Does this PR introduce _any_ user-facing change?
<!--
Note that it means *any* user-facing change including all aspects such as the documentation fix.
If yes, please clarify the previous behavior and the change this PR proposes - provide the console output, description and/or an example to show the behavior difference if possible.
If possible, please also clarify if this is a user-facing change compared to the released Spark versions or within the unreleased branches such as master.
If no, write 'No'.
-->
Now users can pass `scale` parameter to the `floor` and `ceil` functions.
 ```
     > SELECT floor(-0.1);
       -1.0
      > SELECT floor(5);
       5
      > SELECT floor(3.1411, 3);
       3.141
      > SELECT floor(3.1411, -3);
       1000.0

      > SELECT ceil(-0.1);
       0.0
      > SELECT ceil(5);
       5
      > SELECT ceil(3.1411, 3);
       3.142
      > SELECT ceil(3.1411, -3);
       1000.0

```
### How was this patch tested?
<!--
If tests were added, say they were added here. Please make sure to add some test cases that check the changes thoroughly including negative and positive cases if possible.
If it was tested in a way different from regular unit tests, please clarify how you tested step by step, ideally copy and paste-able, so that other reviewers can test and check, and descendants can verify in the future.
If tests were not added, please describe why they were not added and/or why it was difficult to add.
If benchmark tests were added, please run the benchmarks in GitHub Actions for the consistent environment, and the instructions could accord to: https://spark.apache.org/developer-tools.html#github-workflow-benchmarks.
-->
This patch was tested locally using unit test and git workflow.

sathiyapk committed:
[SPARK-37475][SQL] Adds scale parameeter to floor and ceil functions

sathiyapk committed:
[SPARK-37475][SQL] Adds scale parameeter to floor and ceil functions

AmplabJenkins commented:
Can one of the admins verify this patch?

sathiyapk committed:
[SPARK-37475][SQL] Adds scale parameeter to floor and ceil functions

sathiyapk committed:
[SPARK-37475][SQL] Adds scale parameeter to floor and ceil functions

gengliangwang commented on sql/core/src/test/resources/sql-tests/results/operators.sql.out:
```diff
@@ -357,87 +357,87 @@ struct<COT(-1):double>
 select ceiling(0)
 -- !query schema
-struct<ceiling(0):bigint>
+struct<ceiling(0, 0):int>
```
Not every DBMS system supports the second parameter. Some of the external connectors are relying on the SQL representation string of expressions.
So, shall we hide the 2nd parameter if it is the default 0?

sathiyapk replied:
I'm not sure how to do it. Could you please show me an example in the code ? thanks.

cloud-fan replied:
You can follow `Rand`, which needs to hide the seed if not specified.

sathiyapk replied:
ok, thanks !

gengliangwang commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/mathExpressions.scala:
```diff
@@ -453,40 +439,26 @@ case class Expm1(child: Expression) extends UnaryMathExpression(StrictMath.expm1
+      > SELECT _FUNC_(3.1411, 3);
+       3.141
+      > SELECT _FUNC_(3.1411, -3);
+       1000.0
```
why is it 1000?

sathiyapk replied:
If scale_expr is negative, then it specifies the number of places before the decimal point to which to adjust the number. For example, if the scale is -2, then the result is a multiple of 100.

REF: https://docs.snowflake.com/en/sql-reference/functions/ceil.html

sadikovi commented:
@sathiyapk Can you update the PR description? There is a typo, should be `[SPARK-37475][SQL] Add scale parameter to floor and ceil functions`

sathiyapk commented on sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/MathExpressionsSuite.scala:
```diff
@@ -658,6 +658,38 @@ class MathExpressionsSuite extends SparkFunSuite with ExpressionEvalHelper {
+
+    val shortResultsCeil: Seq[Short] = Seq[Short](16960, -31072, -25536, 32000, 31500, 31420) ++
+      Seq.fill[Short](7)(31415)
+
```
Could anybody double check the results of Ceil for shortPI, please ?

sathiyapk commented on sql/core/src/test/resources/sql-tests/results/postgreSQL/float8.sql.out:
```diff
@@ -319,37 +319,37 @@ struct<five:string,f1:double,round_f1:double>
+0.0
+1.0
+1.2345678901234E200
+1005.0
```
In order to avoid any breaking changes, shall we cast the output type of `ceil`/`floor` to Long if the scale is 0 ?

sathiyapk commented:
```
19:40:16.643 ERROR org.apache.spark.sql.SQLQueryTestSuite: Error using configs: 
[info] - describe-table-column.sql *** FAILED *** (1 second, 385 milliseconds)
[info]   describe-table-column.sql
[info]   Expected "...count	1
[info]   avg_col_len	[5]
[info]   max_col_len	5
[info]   histo...", but got "...count	1
[info]   avg_col_len	[4617315517961601024]
[info]   max_col_len	5
[info]   histo..." Result did not match for query #28
[info]   DESC EXTENDED customer cname (SQLQueryTestSuite.scala:473)
```

I'm not sure why the avg_col_len output to [4617315517961601024] instead of [5] in describe-table-column.sql Could anybody help me to figure out, please ? thanks.

cloud-fan commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/mathExpressions.scala:
```diff
@@ -243,40 +243,26 @@ case class Cbrt(child: Expression) extends UnaryMathExpression(math.cbrt, "CBRT"
     Examples:
       > SELECT _FUNC_(-0.1);
-       0
+       0.0
```
Why do we have this behavior change?

sathiyapk replied:
The behavior is same as the `round`/`bround` the output type will be same as the input type.

cloud-fan replied:
For backward compatibility, it will be great if we can keep the previous behavior if the new scale parameter is not specified.

sathiyapk commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/mathExpressions.scala:
```diff
@@ -243,40 +243,26 @@ case class Cbrt(child: Expression) extends UnaryMathExpression(math.cbrt, "CBRT"
       > SELECT _FUNC_(-0.1);
-       0
+       0.0
       > SELECT _FUNC_(5);
```
For an integral type with scale parameter 0, we will get an integral type as the output. We cover this case in the unit test `checkEvaluation(Ceil(5, 0), 5)`. 

We could keep the output type to bigInt if the scale parameter is 0 for `floor`/`ceil` but the behavior will be different than that of `round`/`bround`.

sathiyapk replied:
@cloud-fan @gengliangwang  For the Float and Double type during ROUND_CEILING and ROUND_FLOOR with scale 0, shall we cast the output type to Long ? This will preserve the default behavior of `floor`/`ceil`. Please let me know your opinion.

something like this :
```
  def castOutputToLong: Boolean = Seq(ROUND_CEILING, ROUND_FLOOR).contains(mode) && _scale == 0

     ...
  def nullSafeEval(input1: Any): Any = {
    dataType match {
      ...

      case FloatType =>
        val f = input1.asInstanceOf[Float]
        val output = if (f.isNaN || f.isInfinite) {
          f
        } else {
          BigDecimal(f.toDouble).setScale(_scale, mode).toLong
        }
        if(castOutputToLong) output.toLong else output.toShort

      case DoubleType =>
        val d = input1.asInstanceOf[Double]
        val output = if (d.isNaN || d.isInfinite) {
          d
        } else {
          BigDecimal(d).setScale(_scale, mode).toLong
        }
        if(castOutputToLong) output.toLong else output.toShort
    }
  }
```

sathiyapk committed:
SPARK-37475 Adds scale parameeter to floor and ceil functions

sathiyapk committed:
SPARK-37475 Adds scale parameeter to floor and ceil functions

sathiyapk committed:
SPARK-37475 Reverts floor and ceil STRUCT_TYPE

sathiyapk committed:
SPARK-37475 Creates RoundCeil and RoundFloor case classes that extends RoundBase to handle scale parametes for Ceil and Floor

sathiyapk committed:
SPARK-37475 Creates RoundCeil and RoundFloor case classes that extends RoundBase to handle scale parametes for Ceil and Floor

sathiyapk committed:
SPARK-37475 Creates RoundCeil and RoundFloor case classes that extends RoundBase to handle scale parametes for Ceil and Floor

cloud-fan commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/mathExpressions.scala:
```diff
@@ -249,9 +249,9 @@ case class Cbrt(child: Expression) extends UnaryMathExpression(math.cbrt, "CBRT"
   group = "math_funcs")
-case class Ceil(child: Expression) extends UnaryMathExpression(math.ceil, "CEIL") {
+ case class Ceil(child: Expression) extends UnaryMathExpression(math.ceil, "CEIL") {
   override def dataType: DataType = child.dataType match {
```
Let's define the return type we want:
1. if input is decimal type, we should follow `RoundBase` to define the return type
2. if input is integral type, returning `LongType` as before should be good.
3. if input is float/double, returning `LongType` is definitely wrong. I think returning `DoubleType` should be good.

The problem is what to do if the `scala` parameter is not given. Shall we keep backward compatibility and use the same return type as before? Or we prefer consistency within the system and change the return type?

It looks weird if `ceil(c_double)` and `ceil(c_double, 0)` have different data type. Another idea is to only accept integer constant as the scale parameter, then we can make `ceil` return long type for float/double input if `scale <= 0`.

cc @maropu @viirya @srielau

sathiyapk replied:
@cloud-fan thanks for your review. In the current implementation, if the `scale` parameter is 0 or not specified, we follow the default behaviour of the `Ceil`, in which for the input type float/double, we return `LongType`. If the `scale` parameter is specified and greater than 0, we follow the `RoundBase` behaviour.  I believe `ceil(c_double)` and `ceil(c_double, 0)` will have the same behaviour.

cloud-fan replied:
OK let's only accept int constants for the scale parameter, so that we can keep the return type for `scale <= 0`. We probably need to implement `ExpressionBuilder`:
```
object CeilExpressionBuilder extends ExpressionBuilder {
  def build(expressions: Seq[Expression]): Expression = {
    if (expressions.length == 1) Ceil(expressions.head, 0, LongType)
    else if (expressions.length == 2 && !expressions.last.foldable) {
       fail...
    } else if ...
  }
}

case class Ceil(child: Expression, scale: Int, dataType: DataType)
```

viirya replied:
Do you think will it possibly that `scale` to be not only a literal in the future? Currently looks like the way to keep `ceil(c_double)` and `ceil(c_double, 0)` have consistent data type is to check `scale` value before constructing `Ceil`/`RoundCeil` expr. Seems if `scale` isn't a literal, we cannot do this. If we are pretty sure `scale` cannot be non-literal in the future, then it is fine.

cloud-fan replied:
If we need to support non-literal scale parameter in the future, we just make this behavior change and let `ceil(c_coule, scale)` always return double. I don't see an issue with delaying this behavior change, or never make this behavior change if literal scale parameter is fine.

sathiyapk replied:
@cloud-fan You are right, we need `ExpressionBuilder` for the sql part. I've added the expression builder for `ceil` and `floor` and enough test cases in `MathFunctionsSuite` and it looks fine. I'll push the modifications along with other review comments.

```
object CeilExpressionBuilder extends ExpressionBuilder {
  def build(expressions: Seq[Expression]): Expression = {
    if (expressions.length == 1) Ceil(expressions.head)
    else if (expressions.length == 2) Ceil(expressions(0), expressions(1))
    else throw new AnalysisException("Function ceil cannot take more than 2 parameters")
  }
}

object FloorExpressionBuilder extends ExpressionBuilder {
  def build(expressions: Seq[Expression]): Expression = {
    if (expressions.length == 1) Floor(expressions.head)
    else if (expressions.length == 2) Floor(expressions(0), expressions(1))
    else throw new AnalysisException("Function floor cannot take more than 2 parameters")
  }
}

```

And I believe, for now it will be safer to conserve the same behaviour for `ceil`/`floor` without `scale` parameter or when `scale = 0` and follow the behaviour of `round` when scale parameter is specified and `>0`. I think you prefer the same ?

cloud-fan replied:
We have 2 choices:
1. return double if scale > 0 and return long if scale <= 0. Only accept int literals as the scale parameter, null is also forbidden.
2. always return double, and return null if scale is null.

Can we pick option 1 first as it's stricter? We can always go to option 2 later if it's necessary.

sathiyapk replied:
I agree, for scale <=0 we can always return Long, but for scale > 0 we always return double even when the input type is `Long`/`Int` ? It's better to follow the same behaviour as `round` returning the same output type as the input type for scale > 0, no ?

cloud-fan replied:
Sorry I was not very clear. I was talking about `ceil` function with float/double input. For integral input, it should definitely return long type always.

sathiyapk replied:
@cloud-fan I think we can do something like this, what do you think ?

```
object Ceil {
  def apply(child: Expression, scale: Expression): Expression = {
    val scaleV = scale.eval(EmptyRow)
    if (scaleV == null) return null

    (child.dataType, scaleV.asInstanceOf[Int]) match {
      case (_, 0) => Ceil(child)
      case (_, x) if x < 0 => RoundCeil(Cast(child, LongType), scale)
      case (_: IntegralType | ShortType, _) => RoundCeil(Cast(child, LongType), scale)
      case (_ : FloatType, _) => RoundCeil(Cast(child, DoubleType), scale)
      case _ => RoundCeil(child, scale)
    }
  }
}
```

sathiyapk replied:
Could we can do the same using `DecimalType.bounded(..)`/ `DecimalType.Fixed(..)` ?

cloud-fan replied:
> if (scaleV == null) return null

We can't do it because we need to return an expression which must define a data type.

sathiyapk replied:
I thought the same, but the `eval` of RoundBase does it.

```
override def eval(input: InternalRow): Any = {
    if (scaleV == null) { // if scale is null, no need to eval its child at all
      null
    } else {
      val evalE = child.eval(input)
      if (evalE == null) {
        null
      } else {
        nullSafeEval(evalE)
      }
    }
  }
```

So if we must define the data type of the return expression, we could do something like : 
```
if (scaleV == null) return Literal.create(null, dtNullScale(child.dataType))

def dtNullScale(dt: DataType): DataType = dt match {
  case _ : FloatType | _ : DoubleType => DoubleType
  case _ => LongType
}
```

What do you think ?

cloud-fan replied:
It doesn't matter how we implement it. The problem is we don't have a clear answer for the return type of `ceil(c_double, null)`. `ceil(c_double, 1)` returns double, `ceil(c_double, -1)` returns long, I don't think either long or double makes sense for `ceil(c_double, null)`.

sathiyapk replied:
Or we could let `RoundBase` to manage null value for scale parameter as it does for `round`/`bround`.

sathiyapk replied:
For me personally, returning `DoubleType` for `ceil(c_double, null)` don't shock me, but returning `LongType` for `ceil(c_double, null)` might probably do. And the same goes for `ceil(c_long, null)`.

But for the moment, i think we could either 
1. Let `RoundBase` to deal with the scale will null value. or
2. Throw `AnalysisException` if the scale is null. something like : 
`if (scaleV == null) throw new AnalysisException("Scale parameter can not be null")`

cloud-fan replied:
I'd prefer option 2. This allows us to change the behavior later without a breaking change.

sathiyapk replied:
I agree with you. I'm adding few unit tests to verify the schema of the RoundCeil and RoundFloor. Is there any other tests or changes we may need to add to complete this feature ? I'll push the changes all together. Thanks !

cloud-fan commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/mathExpressions.scala:
```diff
@@ -275,10 +275,27 @@ case class Ceil(child: Expression) extends UnaryMathExpression(math.ceil, "CEIL"
 
+object Ceil {
+  def apply(child: Expression, scale: Expression): Expression = {
+    if (scale.eval(EmptyRow).asInstanceOf[Int] == 0) {
```
This throws exception if `scale` is not literal, right?

sathiyapk replied:
yes, i believe.

viirya replied:
Can `scale` possibly be evaluated to null and get a NPE here?

cloud-fan replied:
ah this is a good point. We should return null if scale is null but the question is which return type? This indicates that we should make the behavior change now and let `ceil(c_double, scala)` always return double type. Let's mention this behavior change in `docs/sql-migration-guide.md`

sathiyapk commented:
@cloud-fan @gengliangwang @sarutak @HyukjinKwon Do you have any comments regarding this PR, please ? I think i need to add few unit-tests regarding the sql struct-type with 2 parameters, i'll do it along with your review comments. Thank you.

viirya commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/mathExpressions.scala:
```diff
@@ -249,9 +249,9 @@ case class Cbrt(child: Expression) extends UnaryMathExpression(math.cbrt, "CBRT"
+ case class Ceil(child: Expression) extends UnaryMathExpression(math.ceil, "CEIL") {
   override def dataType: DataType = child.dataType match {
-    case dt @ DecimalType.Fixed(_, 0) => dt
+    case dt@DecimalType.Fixed(_, 0) => dt
```
We should keep original style.

viirya commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/mathExpressions.scala:
```diff
@@ -249,9 +249,9 @@ case class Cbrt(child: Expression) extends UnaryMathExpression(math.cbrt, "CBRT"
   since = "1.4.0",
   group = "math_funcs")
-case class Ceil(child: Expression) extends UnaryMathExpression(math.ceil, "CEIL") {
+ case class Ceil(child: Expression) extends UnaryMathExpression(math.ceil, "CEIL") {
```
Seems there is redundant space?

viirya commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/mathExpressions.scala:
```diff
@@ -275,10 +275,27 @@ case class Ceil(child: Expression) extends UnaryMathExpression(math.ceil, "CEIL"
       case _ => defineCodeGen(ctx, ev, c => s"(long)(java.lang.Math.${funcName}($c))")
     }
   }
-
```
Unnecessary change.

viirya commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/mathExpressions.scala:
```diff
@@ -457,11 +474,11 @@ case class Expm1(child: Expression) extends UnaryMathExpression(StrictMath.expm1
        5
   """,
-  since = "1.4.0",
+  since = "3.3.0",
```
I think we don't need to change this.

viirya commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/mathExpressions.scala:
```diff
@@ -457,11 +474,11 @@ case class Expm1(child: Expression) extends UnaryMathExpression(StrictMath.expm1
 case class Floor(child: Expression) extends UnaryMathExpression(math.floor, "FLOOR") {
   override def dataType: DataType = child.dataType match {
-    case dt @ DecimalType.Fixed(_, 0) => dt
+    case dt@DecimalType.Fixed(_, 0) => dt
```
ditto

viirya commented on sql/core/src/main/scala/org/apache/spark/sql/functions.scala:
```diff
@@ -1875,15 +1883,23 @@ object functions {
 
   /**
-   * Computes the floor of the given value.
+//   * Computes the floor of the given value of `e` to `scale` decimal places.
+//   *
+//   * @group math_funcs
+//   * @since 3.3.0
+//   */
```
The comment style looks incorrect.

viirya reviewed (commented):
We may need some end-to-end tests too. BTW, this doesn't want to add SQL support, right?

sathiyapk commented:
@viirya Thanks for your comments. I've added `ExpressionBuilder` for `ceil`/`floor` and added enough test cases in `MathFunctionsSuite`. I think we can add some unit tests to verify StructType/Schema for `ceil`/`floor` with scale parameters by adding `xxx.sql` and `xxx.sql.out` files in `sql-tests/results`. 

Do you think we can add any other tests ? If so, could you please show me an example in the source code ? Thank you.

sathiyapk committed:
SPARK-37475 Creates RoundCeil and RoundFloor case classes that extends RoundBase to handle scale parametes for Ceil and Floor

sathiyapk committed:
SPARK-37475 Creates RoundCeil and RoundFloor case classes that extends RoundBase to handle scale parametes for Ceil and Floor

sathiyapk committed:
SPARK-37475 Creates RoundCeil and RoundFloor case classes that extends RoundBase to handle scale parametes for Ceil and Floor

sathiyapk committed:
SPARK-37475 Creates RoundCeil and RoundFloor case classes that extends RoundBase to handle scale parametes for Ceil and Floor

sathiyapk committed:
SPARK-37475 Creates RoundCeil and RoundFloor case classes that extends RoundBase to handle scale parametes for Ceil and Floor

sathiyapk committed:
SPARK-37475 Creates RoundCeil and RoundFloor case classes that extends RoundBase to handle scale parametes for Ceil and Floor

cloud-fan commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/FunctionRegistry.scala:
```diff
@@ -360,6 +360,8 @@ object FunctionRegistry {
     expression[Cbrt]("cbrt"),
     expression[Ceil]("ceil"),
     expression[Ceil]("ceiling", true),
```
We can remove the above two lines now.

cloud-fan commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/FunctionRegistry.scala:
```diff
@@ -369,6 +371,7 @@ object FunctionRegistry {
     expression[Exp]("exp"),
     expression[Expm1]("expm1"),
     expression[Floor]("floor"),
```
ditto

cloud-fan commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/mathExpressions.scala:
```diff
@@ -279,6 +280,49 @@ case class Ceil(child: Expression) extends UnaryMathExpression(math.ceil, "CEIL"
+       0
+      > SELECT _FUNC_(5);
+       5
+      > SELECT _FUNC_(3.1411, 3);
```
can we also show an example with negative scale?

sathiyapk replied:
yes

cloud-fan commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/mathExpressions.scala:
```diff
@@ -279,6 +280,49 @@ case class Ceil(child: Expression) extends UnaryMathExpression(math.ceil, "CEIL"
 }
 
+object Ceil {
+  def apply(child: Expression, scale: Expression): Expression = {
```
It looks an overkill to add a new `Ceil` object. Can we simply put the logic in `CeilExpressionBuilder`?

sathiyapk replied:
This `Ceil` object serves few unit tests in `MathExpressionsSuite`. I am not sure, we can do the same with only `CeilExpressionBuilder` :
`checkEvaluation(checkDataTypeAndCast(Ceil(doublePi, scale)), doubleResultsCeil(i), EmptyRow)`
Could we ?

cloud-fan commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/mathExpressions.scala:
```diff
@@ -279,6 +280,49 @@ case class Ceil(child: Expression) extends UnaryMathExpression(math.ceil, "CEIL"
+case class RoundCeil(child: Expression, scale: Expression)
+  extends RoundBase(child, scale, BigDecimal.RoundingMode.CEILING, "ROUND_CEILING")
+    with Serializable with ImplicitCastInputTypes {
+  def this(child: Expression) = this(child, Literal(0))
```
do we need this?

sathiyapk replied:
No, we don't need it. I'm removing it.

cloud-fan commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/mathExpressions.scala:
```diff
@@ -279,6 +280,49 @@ case class Ceil(child: Expression) extends UnaryMathExpression(math.ceil, "CEIL"
+       3.142
+  """,
+  since = "3.3.0",
+  group = "math_funcs")
```
We can remove the doc string in `case class Ceil` now

sathiyapk replied:
yes, i'm removing it.

cloud-fan commented on sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/MathExpressionsSuite.scala:
```diff
@@ -321,19 +322,29 @@ class MathExpressionsSuite extends SparkFunSuite with ExpressionEvalHelper {
-    testUnary(Ceil, (d: Double) => math.ceil(d).toLong)
-    checkConsistencyBetweenInterpretedAndCodegen(Ceil, DoubleType)
+    testUnary(Ceil.apply, (d: Double) => math.ceil(d).toLong)
+//    checkConsistencyBetweenInterpretedAndCodegen(Ceil, DoubleType)
```
we should make sure existing tests pass

sathiyapk replied:
getting ambiguous reference error between `apply(child)` and `apply(child, scale)`. Do you any recommendation to force `apply(child)` here ?

cloud-fan replied:
`testUnary(Ceil(_), ...`

cloud-fan replied:
actually if we remove the new `object Ceil`, the problem is solved?

sathiyapk replied:
Yes, you are right, if we remove the `object Ceil` the problem will be solved.

cloud-fan commented on sql/core/src/main/scala/org/apache/spark/sql/functions.scala:
```diff
@@ -1755,15 +1755,23 @@ object functions {
+   * @group math_funcs
+   * @since 3.3.0
+   */
+  def ceil(e: Column, scale: Column): Column = withExpr { Ceil(e.expr, scale.expr) }
```
We can use `UnresolvedFunction`

sathiyapk replied:
Could you please give an example for the `UnresolvedFunction`, please ?

cloud-fan replied:
`def ceil(e: Column, scale: Column): Column = withExpr { UnresolvedFunction(Seq("ceil"), Seq(e.expr, scale.expr), isDistinct = false) }`

sathiyapk committed:
SPARK-37475 Creates RoundCeil and RoundFloor case classes that extends RoundBase to handle scale parametes for Ceil and Floor

sathiyapk committed:
SPARK-37475 Creates RoundCeil and RoundFloor case classes that extends RoundBase to handle scale parametes for Ceil and Floor

sathiyapk commented:
@cloud-fan Do you have any other comments regarding this PR, please ?

cloud-fan commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/mathExpressions.scala:
```diff
@@ -279,6 +269,50 @@ case class Ceil(child: Expression) extends UnaryMathExpression(math.ceil, "CEIL"
 }
 
+@ExpressionDescription(
+  usage = "_FUNC_(expr) - Returns the smallest integer not smaller than `expr`.",
```
```suggestion
  usage = "_FUNC_(expr[, scale]) - Returns the smallest number after rounding up that is not smaller than `expr`. A optional `scale` parameter can be specified to control the rounding behavior.",
```

cloud-fan commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/mathExpressions.scala:
```diff
@@ -279,6 +269,50 @@ case class Ceil(child: Expression) extends UnaryMathExpression(math.ceil, "CEIL"
+  group = "math_funcs")
+object CeilExpressionBuilder extends ExpressionBuilder {
+  def build(expressions: Seq[Expression]): Expression = {
+    if (expressions.length == 1) Ceil(expressions.head)
```
nit: follow the java style guide
```
if (...) {

} else if (...) {

} else {

}
```

cloud-fan commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/mathExpressions.scala:
```diff
@@ -279,6 +269,50 @@ case class Ceil(child: Expression) extends UnaryMathExpression(math.ceil, "CEIL"
+    else if (expressions.length == 2) {
+      val child = expressions(0)
+      val scale = expressions(1)
+      val scaleV = scale.eval(EmptyRow)
```
We should fail earlier if `scale.foldable` is not true. otherwise we may hit undefined errors here.

cloud-fan commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/mathExpressions.scala:
```diff
@@ -279,6 +269,50 @@ case class Ceil(child: Expression) extends UnaryMathExpression(math.ceil, "CEIL"
+      val scaleV = scale.eval(EmptyRow)
+      if (scaleV == null) throw new AnalysisException("Scale parameter can not be null")
+
+      (child.dataType, scaleV.asInstanceOf[Int]) match {
```
ditto, we should check `scale.dataType` first, otherwise we may hit cast error when doing `asInstanceOf`

cloud-fan commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/mathExpressions.scala:
```diff
@@ -484,9 +507,53 @@ case class Floor(child: Expression) extends UnaryMathExpression(math.floor, "FLO
+}
+
+@ExpressionDescription(
+  usage = "_FUNC_(expr) - Returns the largest integer not greater than `expr`.",
```
```suggestion
  usage = "_FUNC_(expr[, scale]) - Returns the largest number after rounding down that is not greater than `expr`. An optional `scale` parameter can be specified to control the rounding behavior.",
```

cloud-fan commented on sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/MathExpressionsSuite.scala:
```diff
@@ -321,16 +322,26 @@ class MathExpressionsSuite extends SparkFunSuite with ExpressionEvalHelper {
+
   test("ceil") {
-    testUnary(Ceil, (d: Double) => math.ceil(d).toLong)
+    testUnary(Ceil.apply, (d: Double) => math.ceil(d).toLong)
```
do we still need this change now?

sathiyapk replied:
No need, i removed it.

cloud-fan commented on sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/MathExpressionsSuite.scala:
```diff
@@ -351,24 +362,25 @@ class MathExpressionsSuite extends SparkFunSuite with ExpressionEvalHelper {
     checkEvaluation(checkDataTypeAndCast(Ceil(1)), 1L, EmptyRow)
-    checkEvaluation(checkDataTypeAndCast(Ceil(1234567890123456L)), 1234567890123456L, EmptyRow)
+    checkEvaluation(checkDataTypeAndCast(
+      Ceil(1234567890123456L)), 1234567890123456L, EmptyRow)
```
unnecessary change

cloud-fan commented on sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/MathExpressionsSuite.scala:
```diff
@@ -669,19 +738,62 @@ class MathExpressionsSuite extends SparkFunSuite with ExpressionEvalHelper {
       checkEvaluation(BRound(longPi, scale), longResults(i), EmptyRow)
       checkEvaluation(BRound(floatPi, scale), floatResults(i), EmptyRow)
+      checkEvaluation(checkDataTypeAndCast(
+        FloorExpressionBuilder.build(Seq(Literal(doublePi), Literal(scale)))),
```
This is UT, we can test against `RoundCeil` directly.

sathiyapk replied:
No we can't, because the casting for negative scale, int, short and floats are done at `FloorExpressionBuilder `

cloud-fan commented on sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/MathExpressionsSuite.scala:
```diff
@@ -691,6 +803,19 @@ class MathExpressionsSuite extends SparkFunSuite with ExpressionEvalHelper {
+      try {
+        checkEvaluation(checkDataTypeAndCast(RoundFloor(Literal.create(null, dataType),
+          Literal.create(null, IntegerType))), null)
+        checkEvaluation(checkDataTypeAndCast(RoundCeil(Literal.create(null, dataType),
```
it looks weird to test the error behavior in this test case. Can we move it to `ceil-floor-with-scale-param.sql`?

cloud-fan commented on sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/MathExpressionsSuite.scala:
```diff
@@ -658,6 +671,62 @@ class MathExpressionsSuite extends SparkFunSuite with ExpressionEvalHelper {
+      if (i <= 6) results(i).toLong else results(i)
+    }
+
+    def doubleResultsCeil(i: Int): Any = {
```
is it the same as `doubleResults` for `Round`?

sathiyapk replied:
No, here is the `doubleResults` for `Round` :
`val doubleResults: Seq[Double] = Seq(0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 3.0, 3.1, 3.14, 3.142,
      3.1416, 3.14159, 3.141593)`

cloud-fan commented on sql/core/src/test/resources/sql-tests/inputs/ceil-floor-with-scale-param.sql:
```diff
@@ -0,0 +1,21 @@
+SELECT CEIL(-35, -1);
+SELECT CEIL(-0.1, 0);
+SELECT CEIL(5, 0);
+SELECT CEIL(3.14115, -3);
```
let's add some negative cases
```
SELECT CEIL(2.5, null);
SELECT CEIL(2.5, 'a');
SELECT CEIL(2.5, 0, 0);
```

cloud-fan commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/mathExpressions.scala:
```diff
@@ -279,6 +269,50 @@ case class Ceil(child: Expression) extends UnaryMathExpression(math.ceil, "CEIL"
+      (child.dataType, scaleV.asInstanceOf[Int]) match {
+        case (_, 0) => Ceil(child)
+        case (_, x) if x < 0 => RoundCeil(Cast(child, LongType), scale)
+        case (_: IntegralType | ShortType, _) => RoundCeil(Cast(child, LongType), scale)
```
how about `ByteType`?

sathiyapk replied:
Good question. In case of `ByteType` do we need to cast to `Long`/`Double` ?

sathiyapk replied:
Let's cast ByteType to `LongType` ? I don't find any UT for `Round` with ByteType. Do you any specific case in mind, please ?

cloud-fan replied:
casting to long SGTM

sathiyapk replied:
Ok, i'll do it. I'm not sure about the UT cases though, could you give me few cases, please ? thanks !

cloud-fan replied:
actually, you wrote `IntegralType` here, which covers byte/short/int/long already.

sathiyapk replied:
ah ok, so let's remove ShortType and do this ?
```
(child.dataType, scaleV.asInstanceOf[Int]) match {
        case (_, 0) => Ceil(child)
        case (_, x) if x < 0 => RoundCeil(Cast(child, LongType), scale)
        case (_: IntegralType, _) => RoundCeil(Cast(child, LongType), scale)
        case (_ : FloatType, _) => RoundCeil(Cast(child, DoubleType), scale)
        case _ => RoundCeil(child, scale)
}
```

sathiyapk committed:
SPARK-37475 Creates RoundCeil and RoundFloor case classes that extends RoundBase to handle scale parametes for Ceil and Floor

sathiyapk commented:
@cloud-fan Any idea why we getting 
` Failed to execute goal net.alchim31.maven:scala-maven-plugin:4.3.0:doc-jar (attach-scaladocs) on project spark-catalyst_2.12: MavenReportException: Error while creating archive: wrap: Process exited with an error: 1`, please ?

cloud-fan commented on sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/MathExpressionsSuite.scala:
```diff
@@ -357,18 +367,18 @@ class MathExpressionsSuite extends SparkFunSuite with ExpressionEvalHelper {
 
   test("floor") {
-    testUnary(Floor, (d: Double) => math.floor(d).toLong)
+    testUnary(Floor.apply, (d: Double) => math.floor(d).toLong)
```
unnecessary change?

cloud-fan commented on sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/MathExpressionsSuite.scala:
```diff
@@ -357,18 +367,18 @@ class MathExpressionsSuite extends SparkFunSuite with ExpressionEvalHelper {
     checkConsistencyBetweenInterpretedAndCodegen(Floor, DoubleType)
 
-    testUnary(Floor, (d: Decimal) => d.floor, (-20 to 20).map(x => Decimal(x * 0.1)))
+    testUnary(Floor.apply, (d: Decimal) => d.floor, (-20 to 20).map(x => Decimal(x * 0.1)))
```
ditto

cloud-fan commented on sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/MathExpressionsSuite.scala:
```diff
@@ -381,7 +391,8 @@ class MathExpressionsSuite extends SparkFunSuite with ExpressionEvalHelper {
     checkEvaluation(checkDataTypeAndCast(Floor(1)), 1L, EmptyRow)
-    checkEvaluation(checkDataTypeAndCast(Floor(1234567890123456L)), 1234567890123456L, EmptyRow)
+    checkEvaluation(checkDataTypeAndCast(
+      Floor(1234567890123456L)), 1234567890123456L, EmptyRow)
```
ditto

cloud-fan commented on sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/MathExpressionsSuite.scala:
```diff
@@ -658,6 +669,62 @@ class MathExpressionsSuite extends SparkFunSuite with ExpressionEvalHelper {
+      results(i)
+    }
+
+    def shortResultsCeil(i: Int): Any = {
```
Since the data type of `ceil` is always double/long, I think we should only test double and long. It's not useful to test byte/short/int/float, as Spark will cast them to double/long and we don't need to test the cast here.

sathiyapk replied:
I think we need to keep this to test `Ceil` with Scale parameters to verify whether we cast it to right type.

cloud-fan replied:
This test should focus on expression evaluation and we should test `RoundCeil` diretly. If we want to test the added cast, we should add a new test case, to call `Ceil.apply` and check the result expression to see if it has the rigjt cast, without evaluating it.

sathiyapk replied:
Sorry, i am not sure i got it right. 

Actually we test `RoundCeil` and `RoundFloor` directly for positive `scale` on big decimal, i can add long and double in that case, if you prefer. But we can't test `RoundCeil` and `RoundFloor` directly for negative `scale` because the casting to `Long` for negative scale happens one step before. 

May be if put the casting in the `RoundCeil` and `RoundFloor` objects and let the `ExpressionBuilder` call these objects like it was before, we can able to test `RoundCeil` and `RoundFloor` directly for both negative and positive `scale`. What do you say ?

sathiyapk replied:
@cloud-fan could you please add your opinion on this point as well ? Do you agree to put the cast in the `RoundCeil`/`RoundFloor` object to be able to unit test the `RoundCeil` and `RoundFloor` directly for both +ve and -ve scale ?

cloud-fan replied:
Ideally, expression builder should simply be a router and create corresponding expressions for certain input types. The type coercion shouldn't be done there. The created expression should define its expected input types and let the type coercion framework to add casts.

sathiyapk replied:
Ok perfect, i'll put back the casting in the `RoundCeil` and `RoundFloor` objects then.

cloud-fan commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/mathExpressions.scala:
```diff
@@ -279,6 +269,59 @@ case class Ceil(child: Expression) extends UnaryMathExpression(math.ceil, "CEIL"
+      val child = expressions(0)
+      val scale = expressions(1)
+      if (! (scale.foldable && scale.dataType == DataTypes.IntegerType)) {
+        throw new AnalysisException("Invalid scale parameter for the function ceil")
```
let's define the error in `QueryCompilationError.scala`

cloud-fan replied:
The error message can be: `The 'scale' parameter of function ceil must be an int constant.`

sathiyapk committed:
SPARK-37475 Creates RoundCeil and RoundFloor case classes that extends RoundBase to handle scale parametes for Ceil and Floor

cloud-fan commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/mathExpressions.scala:
```diff
@@ -279,6 +269,59 @@ case class Ceil(child: Expression) extends UnaryMathExpression(math.ceil, "CEIL"
 
+@ExpressionDescription(
+  usage = """
+  _FUNC_(expr[, scale]) - Returns the smallest number after rounding up that is not smaller
```
nit: the usage string should not have new lines in it. We can follow `TryAdd` to define very long usage string.

cloud-fan commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/mathExpressions.scala:
```diff
@@ -484,9 +516,62 @@ case class Floor(child: Expression) extends UnaryMathExpression(math.floor, "FLO
+
+@ExpressionDescription(
+  usage = """
+  _FUNC_(expr[, scale]) - Returns the largest number after rounding down that is not greater
```
ditto

cloud-fan commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/mathExpressions.scala:
```diff
@@ -484,9 +516,62 @@ case class Floor(child: Expression) extends UnaryMathExpression(math.floor, "FLO
+  """,
+  since = "3.3.0",
+  group = "math_funcs")
+object FloorExpressionBuilder extends ExpressionBuilder {
```
This is very similar to `CeilExpressionBuilder`. Can we have a `trait CeilFloorExprBuilderBase` to put the common code?

cloud-fan commented on sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/MathExpressionsSuite.scala:
```diff
@@ -705,6 +818,42 @@ class MathExpressionsSuite extends SparkFunSuite with ExpressionEvalHelper {
+    checkEvaluation(checkDataTypeAndCast(
+      CeilExpressionBuilder.build(Seq(Literal(5), Literal(0)))), 5L)
+    checkEvaluation(checkDataTypeAndCast(
+      CeilExpressionBuilder.build(Seq(Literal(3.1411), Literal(-3)))), 1000L)
```
I'm a bit concerned about this case. It means the result can be arbitrarily big, if the `scale` is a very large negative number.

Since we return long for negative scale, I think the scale can't be smaller than `-18`. However, this can be weird if the input is decimal, e.g. there is a column `col` of type `decimal(38, 0)`, it's weird if we fail `ceil(col, -30)`.

It seems to be that, `RoundCeil` should only accept decimal inputs if the scale is not 0. Then return type should consider both the input decimal type and the scale value.

sathiyapk replied:
That's a good point. 

you wanted to say `RoundCeil should only accept decimal inputs if the scale is` **-ve** ? RoundCeil can accept decimal inputs if the scale is +ve, right ?

And this specific case don't apply to `RoundFloor` right ?

cloud-fan replied:
`RoundFloor` has a similar issue, if the input is a negative value.

My proposal is, for `round/ceil(input, scale)` if `scale != 0`, the expected input type should be decimal type, and the type coercion framework should cast int, double, etc. to decimal type. The result type can be
```
case DecimalType.Fixed(p, s) =>
  if (scale < 0) {
    DecimalType(math.max(p - s, 1 - scala), 0)
  } else {
    DecimalType(p, math.min(s, scale))
  }
```

sathiyapk replied:
@cloud-fan Do you think something like this would work ?

```
(child.dataType, scaleV.asInstanceOf[Int]) match {
    case (DecimalType.Fixed(p, s), x) if x < 0 =>
      val castedChild = Cast(child, DecimalType(math.max(p, 1 - x), 0))
      RoundCeil(castedChild, scale)
    case (DecimalType.Fixed(p, s), x) =>
      val castedChild = Cast(child, DecimalType(p, math.min(s, x)))
      RoundCeil(castedChild, scale)
    case _ => RoundCeil(child, scale)
}
```

cloud-fan replied:
Something similar, but we shouldn't add cast on our own. We should rely on the framework
```
if (scaleV.asInstanceOf[Int] == 0) {
  Ceil(child)
} else {
  RoundCeil(child, scale)
}
...
case class RoundCeil ... {
  def inputTypes = Seq(DecimalType, IntegerType)
  def dataType = child.dataType match {
    case DecimalType.Fixed(p, s) =>
      if (scale < 0) {
        DecimalType(math.max(p, 1 - scale), 0)
      } else {
        DecimalType(p, math.min(s, scale))
      }
  }
}
```

sathiyapk replied:
@cloud-fan So basically if the scale is `scale < 0`, this produces the result of `scale = 0` and the return type is always decimal. This don't shock me, but this will be different from the behaviour of the snow-flake `ceil`/`floor` function. Are you sure about this ?

REF: https://docs.snowflake.com/en/sql-reference/functions/ceil.html
`If scale_expr is negative, then it specifies the number of places before the decimal point to which to adjust the number. For example, if the scale is -2, then the result is a multiple of 100.`

cloud-fan replied:
> this produces the result of scale = 0 and the return type is always decimal.

No, I'm not saying we should change the calculation. I mean we should return decimal type, so that it can be bigger than int64 (decimal type can hold 38 digits).

sathiyapk replied:
@cloud-fan with `def inputTypes = Seq(DecimalType, IntegerType)`, the framework will cast all inputs to DecimalType and `-ve` scale for decimals are not allowed by default. If we put negative scale for decimal, we will receive the following error: 
`Negative scale is not allowed: -2. You can use spark.sql.legacy.allowNegativeScaleOfDecimal=true to enable legacy mode to allow it.`

In the code actuel, i think it would be enough to add `math.min(scale, -18)` for the `scale < 0` something as bellow.

(child.dataType, scaleV.asInstanceOf[Int]) match {
        case (_, 0) => Ceil(child)
        case (_, x) if x < 0 => RoundCeil(Cast(child, LongType), math.min(-18, scale))
        case (_: IntegralType, _) => RoundCeil(Cast(child, LongType), scale)
        case (_ : FloatType, _) => RoundCeil(Cast(child, DoubleType), scale)
        case _ => RoundCeil(child, scale)
      }

What do you think ?

cloud-fan replied:
In my previous comment, I already proposed the way to deal with negative scale
```
case DecimalType.Fixed(p, s) =>
  if (scale < 0) {
    DecimalType(math.max(p, 1 - scale), 0)
  } else {
    DecimalType(p, math.min(s, scale))
  }
```

sathiyapk replied:
@cloud-fan Here is the code snippet that evaluates DecimalType at `RoundBase`. Basically if we assign `s = 0` for the negative scale, then for `scale < 0`, it will produce the result of `scale = 0`.
```
def nullSafeEval(input1: Any): Any = {
    dataType match {
      case DecimalType.Fixed(_, s) =>
        val decimal = input1.asInstanceOf[Decimal]
        // Overflow cannot happen, so no need to control nullOnOverflow
        decimal.toPrecision(decimal.precision, s, mode)
...
}
```

I couldn't find any simpler way to deal negative scale with `def inputTypes = Seq(DecimalType, IntegerType)`. Do you have any other proposals to deal negative scale, please ? thanks.

sathiyapk replied:
@cloud-fan do you have any update on this please ?

cloud-fan replied:
It's easy to fix `round`, as it returns 0 if the negative scale is too big. e.g. `round(12.3, -3)` returns 0. But `ceil/floor` is more complicated as it may need to increase precision for large negative scale parameters.

My previous proposal is actually correct
```
def dataType ...
  case DecimalType.Fixed(p, s) =>
    if (scale < 0) {
      DecimalType(math.max(p, 1 - scale), 0)
    } else {
      DecimalType(p, math.min(s, scale))
    }
```
But we need to fix the eval function as well
```
def nullSafeEval(input1: Any): Any = {
    dataType match {
      case DecimalType.Fixed(p, s) =>
        val decimal = input1.asInstanceOf[Decimal]
        if (_scale >=0) {
          // Overflow cannot happen, so no need to control nullOnOverflow
          decimal.toPrecision(decimal.precision, s, mode)
        } else {
          Decimal(decimal.toBigDecimal.setScale(_scale, mode), p, s)
        }
...
}
```

sathiyapk replied:
@cloud-fan Perfect, that does the work. Could you please confirm the `doGenCode` part below, please ?

```
case DecimalType.Fixed(_, s) =>
        s"""
          if (${_scale} >= 0) {
            ${ev.value} = ${ce.value}.toPrecision(${ce.value}.precision(), $s,
            Decimal.$modeStr(), true);
            ${ev.isNull} = ${ev.value} == null;
          } else {
            ${ev.value} = new java.math.BigDecimal(${ce.value}).
            setScale(${_scale}, java.math.BigDecimal.${modeStr});
            ${ev.isNull} = ${ev.value} == null;
          }"""
```

cloud-fan replied:
The codegen can be smarter, because we know the result of `_scale >= 0` before generating code
```
if (_scale >= 0) {
  "java code"
} else {
  "java code"
}
```

sathiyapk replied:
@cloud-fan Could you please give me the "java code" for the else part, please ? I'm getting compile exception on 'generated.java'.

cloud-fan replied:
Can you push whatever you have first? Then I can edit the file directly with GitHub.

sathiyapk replied:
@cloud-fan I pushed the code 

```
override def doGenCode(ctx: CodegenContext, ev: ExprCode): ExprCode = {
    val ce = child.genCode(ctx)

    val evaluationCode = dataType match {
      case DecimalType.Fixed(p, s) =>
        if (_scale >= 0) {
          s"""
            ${ev.value} = ${ce.value}.toPrecision(${ce.value}.precision(), $s,
            Decimal.$modeStr(), true);
            ${ev.isNull} = ${ev.value} == null;"""
       } else {
          s"""
            ${ev.value} = ${ce.value}.toPrecision(${ce.value}.precision(), $s,
            Decimal.$modeStr(), true);
            ${ev.isNull} = ${ev.value} == null;"""
//          s"""
//          ${ev.value} = new java.math.BigDecimal(${ce.value}).
//            setScale(${_scale}, java.math.BigDecimal.${modeStr});"""
        }
```

Could you please check what's wrong with the following code for the else part, please ? thanks
```
s"""
          ${ev.value} = new java.math.BigDecimal(${ce.value}).
           setScale(${_scale}, java.math.BigDecimal.${modeStr});"""
```

sathiyapk committed:
SPARK-37475 Creates RoundCeil and RoundFloor case classes that extends RoundBase to handle scale parametes for Ceil and Floor

cloud-fan commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/mathExpressions.scala:
```diff
@@ -1426,12 +1526,21 @@ abstract class RoundBase(child: Expression, scale: Expression,
+            Decimal.$modeStr(), true);
+            ${ev.isNull} = ${ev.value} == null;"""
+//          s"""
+//          ${ev.value} = new java.math.BigDecimal(${ce.value}).
+//            setScale(${_scale}, java.math.BigDecimal.${modeStr});"""
```
```suggestion
//          ${ev.value} = new Decimal(new java.math.BigDecimal(${ce.value}).
//            setScale(${_scale}, java.math.BigDecimal.${modeStr}));"""
```

cloud-fan replied:
ah, we should follow the scala code: `new Decimal(${ce.value}. toJavaBigDecimal.setScale...)`

sathiyapk replied:
@cloud-fan I get the following error: 
org.codehaus.commons.compiler.CompileException: File 'generated.java', Line 56, Column 2: **"toJavaBigDecimal" is neither a method, a field, nor a member class of "org.apache.spark.sql.types.Decimal"**

```
s"""
            ${ev.value} = new Decimal(${ce.value}.toJavaBigDecimal
            .setScale(${_scale}, java.math.BigDecimal.${modeStr}));"""
```

cloud-fan replied:
This is java code, so parentheses is required when calling a function. try `... toJavaBigDecimal()...`?

sathiyapk committed:
SPARK-37475 Adds scale parameter to ceil and floor functions

cloud-fan commented:
thanks, merging to master!

sathiyapk commented:
thanks !
